### ISSUE TITLE
recursion in PubNub::PubSub::subscribe causes memory leak

### ISSUE THREAD

**Issue by tofurky** [binary-com/perl-pubnub-pubsub#21]
i noticed that a certain code path in the subscribe subroutine is calling itself in a recursive fashion:

```
    my $rtn = $callback ? $callback->(@cb_args) : 1;
    return unless $rtn;

    $timetoken = $json->[1];
    return $self->subscribe(%params, timetoken => $timetoken);
}
```

over long periods of time, this has caused my script using this module to leak hundreds of MB of memory. to give you an idea of what is leaking, here is some output generated (in part) by using Devel::Gladiator:
```
leak count increased for ARRAY from 7698 to 7720
leak count increased for CODE from 6756 to 6757
leak count increased for HASH from 2695 to 2704
leak count increased for Mojo::Asset::Memory from 12 to 14
leak count increased for Mojo::Content::Single from 12 to 14
leak count increased for Mojo::Headers from 12 to 14
leak count increased for Mojo::Message::Request from 6 to 7
leak count increased for Mojo::Message::Response from 6 to 7
leak count increased for Mojo::Parameters from 6 to 7
leak count increased for Mojo::Path from 6 to 7
leak count increased for Mojo::Transaction::HTTP from 6 to 7
leak count increased for Mojo::URL from 6 to 7
leak count increased for REF from 6858 to 6909
leak count increased for REF-ARRAY from 761 to 781
leak count increased for REF-CODE from 1067 to 1075
leak count increased for REF-HASH from 1343 to 1351
leak count increased for REF-Mojo::Asset::Memory from 12 to 14
leak count increased for REF-Mojo::Content::Single from 12 to 14
leak count increased for REF-Mojo::Headers from 12 to 14
leak count increased for REF-Mojo::Message::Request from 6 to 7
leak count increased for REF-Mojo::Message::Response from 6 to 7
leak count increased for REF-Mojo::Parameters from 6 to 7
leak count increased for REF-Mojo::Path from 6 to 7
leak count increased for REF-Mojo::Transaction::HTTP from 7 to 8
leak count increased for REF-Mojo::URL from 6 to 7
leak count increased for REF-Mojo::UserAgent from 14 to 16
leak count increased for REF-PubNub::PubSub from 9 to 10
leak count increased for SCALAR from 91869 to 92056
```

the fix seems to be relatively simple, and is already being used in another code path in the subscribe subroutine. instead of calling `return $self->subscribe(%params, timetoken => $timetoken);`, setting `@_ = ($self, %params, timetoken => $timetoken);` and instead calling `goto &subscribe; `seems to keep the leak from happening.

**Comment by tofurky** [comment]
i created PR https://github.com/binary-com/perl-pubnub-pubsub/pull/22 to fix this.

**Comment by tom-binary** [comment]
Closing, since this is fixed by #22 - thanks!
